doc/tfsformat.rst

```diff
@@ -0,0 +1,142 @@
+
+Written TFS files follow a structure in two parts: first the ``headers`` lines and then the ``table`` itself.
+They are described below, including their specifics.
+
+Headers
+-------
+
+The first section of the TFS file contains `headers`, lines each containing a parameter, its type and its value.
+All headers lines start with the `@` character, and have the following structure:
+
+- The `@` character, to identify the line as a header.
+- The name of the parameter.
+- The type identifier of the parameter (see `Types and Identifiers`_ below).
+- The value of the parameter.
+
+An example header line is:
+
+.. code-block::
+
+    @ NAME                   %le             0.269975
+
+Additionally, any headers line starting with the `#` character is considered a comment and is ignored by the parser.
+There is no limit on the number of header lines in a TFS file, and there can be no header lines at all.
+
+Table
+-----
+
+After header lines follows the `table` section.
+A first line, starting with the `*` character, contains the names of the columns.
+Just below, a line starting with the `$` character contains the type identifiers of the columns' data (see `Types and Identifiers`_ below).
+These two lines look like:
+
+.. code-block::
+
+    *             INTEGERS              STRINGS               FLOATS
+    $                   %d                   %s                  %le
+
+
+Afterwards all lines contain the data of the table, one row per line and columns separated by at least one blank space.
+For readability the entries are traditionally right-aligned with their column names,
+and the column width is chosen to accomodate the longest column name or value, but neither is mandatory.
+Following the example above, a few table lines would look like:
+
+.. code-block::
+
+    *             INTEGERS              STRINGS               FLOATS
+    $                   %d                   %s                  %le
+                         1            SOMETHING              4.51395
+                     21345         "WITH SPACE"             123.4825
+
+
+Types and Identifiers
+---------------------
+
+Type identifiers are specific strings that indicate as which type to interpret the given data as, in the individual entries of the `Headers`_ as well as for the columns of the `Table`_.
+The following type identifiers and their corresponding Python types are accepted and used by `tfs-pandas`:
+
+================  ======================  =============== =========================
+Type Indentifier  Associated Python Type          Example               Accepted by
+================  ======================  =============== =========================
+%s                                string        `%s name`  ``MAD-X`` and ``MAD-NG``
+%bpm_s                            string    `%bpm_s name`  ``MAD-X`` and ``MAD-NG``
+%d                        64-bit integer         `%d 174`  ``MAD-X`` and ``MAD-NG``
```

Comment:
Looking at MAD-NG source code, I cannot see that %d is parsed in the headers - have you tested this? It doesn't matter in columns as it appears MAD-NG ignores the column type and parses item by item and just seems to convert to whatever it can, in the order number->complex->range->boolean->nil.

A little nuance: MAD-NG uses Lua 5.2, which does not have integers as such (https://www.lua.org/pil/2.3.html), so integers have a max of 52-bits (I believe this is the highest precision before its a float).

Reply:
Oh, interesting. I have not used `MAD-NG` for work, do you think you can quickly try to load the `tests/inputs/madng.tfs` file and see if that goes fine?

Reply:
Good catch. I just tried to have `MAD-NG` read the `tests/inputs/madng.tfs` file and it just cries with "invalid header field type: 'd'". I will have to work around that.